### PR TITLE
Python Boto3 and Amazon DynamoDB Programming Tutorial

### PR DESCRIPTION
**Pull request by Tsanguu**
### Article Checklist
**NOTE: (1st time contributors - submit ONLY one article to get familiar with the process, quality, and style-once published, you can then submit any number of articles.)**
#### Preliminary Checks - Formatting and Structure 
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md)**
- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**
- [x] Have you included a hero image that is 600x400 pixels and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
- [x] In your PR - add a comment that states **["This fixes #(enter your Idea Content Form #)"](https://youtu.be/TKJ4RdhyB5Y?t=278)** to link and close your Idea Content (issue) when the PR is merged.
      
#### Grammar Checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) -  [Grammarly](https://grammarly.com) 
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [x] Have you added sources for quotes and images that aren't yours?

#### Technical Checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**

### Contribution Guidelines
For first-time contributors and for more details, **see [Contributing Guidelines](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md)**

### CODE REVIEW

**Comment by Tsanguu** [comment]
This solves #929

**Comment by hectorkambow** [comment]
@paulodhiambo  

I see that you have assigned this article for review - if you are working on another one - could @rahulbanerjee26  

do a test review on this one - he is interested in becoming a peer reviewer. 

If you could please guide @rahulbanerjee26  if he has any questions along the way. 

Do not hesitate to ping me for anything 🚀  
Thank you

File: articles/python-boto3-and-amazon-dynamodb-programming-tutorial/index.md [hunk 1]
```diff
@@ -0,0 +1,587 @@
...
+
+```
+
+Run the command below to execute the script `read_item.py`.
+
+```bash
+python read_item.py
+```
+
+Below is the expected output. The response item is similar to the item we created previously. Using the specific primary key, we can retrieve a particular item.
```
Comment (rahulbanerjee26) [inline articles/python-boto3-and-amazon-dynamodb-programming-tutorial/index.md:301]:
```suggestion
Below is the expected output. You can confirm that the response item is the item we created previously. Using the specific primary key, we can retrieve a particular item.
```

File: articles/python-boto3-and-amazon-dynamodb-programming-tutorial/index.md [hunk 2]
```diff
@@ -0,0 +1,587 @@
...
+          'temperature3': '12.43',
+          'temperature4': '44.74',
+          'temperature5': '12.74'}}
+```
+
+### Conditions
+DynamoDB has a provision of using conditions. Conditions can be applied when updating or deleting items. We can provide a `ConditionExpression`. If the `ConditionExpression` evaluates to true, then the action is performed. Refer [here](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/Expressions.ConditionExpressions.html) for more information on condition expressions. Familiarize yourself with different [DynamoDB conditions](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/customizations/dynamodb.html#dynamodb-conditions).
+
+### Update
+The update modifies an existing item by updating the values of current attributes, removing attributes, or adding new attributes. In this tutorial, we will update the values of existing attributes. Below is the original item and the updated item.
```
Comment (rahulbanerjee26) [inline articles/python-boto3-and-amazon-dynamodb-programming-tutorial/index.md:319]:
```suggestion
Update refers to modifying a previously created item by updating the values of existing attributes, removing attributes, or adding new attributes. In this tutorial, we will update the values of existing attributes. Below is the original item and the updated item.
```

File: articles/python-boto3-and-amazon-dynamodb-programming-tutorial/index.md [hunk 3]
```diff
@@ -0,0 +1,587 @@
...
+    "info_timestamp": "1612522800",
+    "temperature1": 33.74,
+    "temperature2": 23.74,
+    "temperature3": 25.2,
+    "temperature4": 22.0,
+    "temperature5": 25.0
+  }
+}
+```
+We will use the method `update_item`, as shown in the code below. Create a script named update_item.py `update_item.py` and add the code below.
```
Comment (rahulbanerjee26) [inline articles/python-boto3-and-amazon-dynamodb-programming-tutorial/index.md:352]:
```suggestion
We will use the `update_item` method, as shown in the code below. Create a script named  `update_item.py` and add the code below.
```

File: articles/python-boto3-and-amazon-dynamodb-programming-tutorial/index.md [hunk 4]
```diff
@@ -0,0 +1,587 @@
...
+                                      'server': 'Jetty(9.4.18.v20190429)',
+                                      'x-amz-crc32': '1118861638',
+                                      'x-amzn-requestid': 'a6a8201d-dc10-4837-be6a-7de03ee9b24f'},
+                      'HTTPStatusCode': 200,
+                      'RequestId': 'a6a8201d-dc10-4837-be6a-7de03ee9b24f',
+                      'RetryAttempts': 0}}
+```
+
+### Delete Item
+To delete an item, we use the method `delete_item`. We can specify the primary key for the item to delete or provide a `ConditionExpression`. If we use a `ConditionExpression`, the item will not be deleted unless the condition is fulfilled. In this example, we will provide a primary key for the item to be deleted and provide a `ConditionExpression`. The item will be deleted if the `ConditionExpression` is met. In this example, the condition is:
```
Comment (rahulbanerjee26) [inline articles/python-boto3-and-amazon-dynamodb-programming-tutorial/index.md:416]:
```suggestion
To delete an item, we use the `delete_item` method. We can specify the primary key for the item to delete or provide a `ConditionExpression`. If we use a `ConditionExpression`, the item will not be deleted unless the condition is evaluated to True. In this example, we will provide a primary key for the item to be deleted and provide a `ConditionExpression`. The item will be deleted if the `ConditionExpression` is met. In this example, the condition is:
```

File: articles/python-boto3-and-amazon-dynamodb-programming-tutorial/index.md [hunk 5]
```diff
@@ -0,0 +1,587 @@
...
+    "temperature1": 33.74,
+    "temperature2": 23.74,
+    "temperature3": 25.2,
+    "temperature4": 22.0,
+    "temperature5": 25.0
+  }
+}
+```
+
+The item will be deleted if the value of `info_timestamp` is greater than or equal to the value provided. Create a script named `info_timestamp` and paste the code below.
```
Comment (rahulbanerjee26) [inline articles/python-boto3-and-amazon-dynamodb-programming-tutorial/index.md:438]:
```suggestion
The item will be deleted if the value of `info_timestamp` is greater than or equal to the value provided. Create a script named `info_timestamp.py` and paste the code below.
```

File: articles/python-boto3-and-amazon-dynamodb-programming-tutorial/index.md [hunk 6]
```diff
@@ -0,0 +1,587 @@
...
+```
+
+Run the command below to execute the script `query.py`.
+
+```bash
+python query.py
+```
+
+#### Scan
+Scan operation reads and returns all the items in the table. The method `DynamoDB.Table.scan()` is used to scan. Using a `filter_expression`, we can filter the items to be returned. However, the whole table will be scanned, and items not matching the `filter_expression` will be thrown away. Create a script named `scan.py` and paste the code below. To learn more about DynamoDB scans, refer to the [developer guide](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/Scan.html).
```
Comment (rahulbanerjee26) [inline articles/python-boto3-and-amazon-dynamodb-programming-tutorial/index.md:526]:
```suggestion
Scan operation reads and returns all the items in the table. The method `DynamoDB.Table.scan()` is used to scan a table. Using a `filter_expression`, we can filter the items to be returned. However, the whole table will be scanned, and items not matching the `filter_expression` will be thrown away. Create a script named `scan.py` and paste the code below. To learn more about DynamoDB scans, refer to the [developer guide](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/Scan.html).
```

File: articles/python-boto3-and-amazon-dynamodb-programming-tutorial/index.md [hunk 7]
```diff
@@ -0,0 +1,587 @@
...
+        for device in devices:
+            print(f"\n{device['device_id']} : {device['datacount']}")
+            print(device['info'])
+
+    print(
+        f"Scanning all devices data")
+    scan_devices(print_devices)
+```
+
+The script above scans through the table `Devices` with no `filter_expression`. Run the command below to execute the script scan.py. The output will be all the items in the `Devices` table.
```
Comment (rahulbanerjee26) [inline articles/python-boto3-and-amazon-dynamodb-programming-tutorial/index.md:557]:
```suggestion
The script above scans the `Devices` table with no `filter_expression`. Run the command below to execute the script scan.py. The output will be all the items in the `Devices` table.
```

**Review by rahulbanerjee26 — CHANGES_REQUESTED** [review]
Thanks, @hectorkambow, and @paulodhiambo. 

Hi @Tsanguu, as Hector mentioned, I am a new peer reviewer so please feel free to ignore some of my suggestions.

It is a well-written article and easy to understand. I have made a few suggestions to improve readability. You should also consider adding more comments to your code. It will make it easier for readers to understand.

**Comment by rahulbanerjee26** [comment]
@Tsanguu I will review your new changes and move on to the tech review on Sunday.

**Review by rahulbanerjee26 — CHANGES_REQUESTED** [review]
@Tsanguu  Preliminary and Tech Review are completed. I have addressed a minor issue I found during the tech review. Please address that and we can move on to the next steps in the review process.

File: articles/python-boto3-and-amazon-dynamodb-programming-tutorial/index.md [hunk 8]
```diff
@@ -0,0 +1,601 @@
...
+                'datacount': datacount
+            },
+            # Conditional request
+            ConditionExpression="info.info_timestamp <= :value",
+            ExpressionAttributeValues={
+                ":value": info_timestamp
+            }
+        )
+    except ClientError as er:
+        if er.response1['Error']['Code'] == "ConditionalCheckFailedException":
```
[the remark this answers is not in the document]
Comment (Tsanguu) [reply articles/python-boto3-and-amazon-dynamodb-programming-tutorial/index.md]: @rahulbanerjee26 I have taken note of the issues and resolved the code error.

**Comment by rahulbanerjee26** [comment]
@hectorkambow The preliminary review and tech review for this article are completed. It requires a plagiarism check and is ready for the final review.

**Review by rahulbanerjee26 — APPROVED** [review]

**Comment by hectorkambow** [comment]
plagiarism check done

**Comment by hectorkambow** [comment]
Great work gentlemen

@Tsanguu  

cc @rahulbanerjee26  🚀

**Pull request merged**